Userland/Libraries/LibJS/AST.h

```diff
@@ -97,6 +97,45 @@ class ASTNode : public RefCounted<ASTNode> {
     u32 m_end_offset { 0 };
 };
 
+// This is a helper class that packs an array of T after the AST node, all in the same allocation.
+template<typename Derived, typename Base, typename T>
+class ASTNodeWithTailArray : public Base {
+public:
+    template<typename... Args>
+    static NonnullRefPtr<Derived> create(size_t tail_size, SourceRange source_range, Args&&... args)
+    {
+        static_assert(alignof(Derived) % alignof(T) == 0, "need padding for tail array");
+        auto memory = ::operator new(sizeof(Derived) + tail_size * sizeof(T));
+        return adopt_ref(*::new (memory) Derived(move(source_range), static_cast<decltype(args)>(args)...));
+    }
+
+    virtual ~ASTNodeWithTailArray() override
+    {
+        for (auto& value : tail_span())
+            value.~T();
+    }
+
+    ASTNodeWithTailArray(SourceRange source_range, Span<T const> values)
```

Comment:
I wouldn't recommend making this constructor public. You don't want to accidentally create it without the tail allocation.

Reply:
Good point!